geonode/thumbs/utils.py

```diff
@@ -193,74 +185,63 @@ def construct_wms_url(
         thumbnail_url = ogc_server_settings.LOCATION
 
     wms_endpoint = ""
+    additional_kwargs = {}
     if thumbnail_url == ogc_server_settings.LOCATION:
         # add access token to requests to Geoserver (logic based on the previous implementation)
         username = ogc_server_settings.credentials.username
         user = get_user_model().objects.filter(username=username).first()
         if user:
             access_token = get_or_create_token(user)
             if access_token and not access_token.is_expired():
-                params["access_token"] = access_token.token
+                additional_kwargs['access_token'] = access_token.token
 
         # add WMS endpoint to requests to Geoserver
         wms_endpoint = getattr(ogc_server_settings, "WMS_ENDPOINT") or "ows"
 
-    thumbnail_url = f"{thumbnail_url}{wms_endpoint}?{'&'.join(f'{key}={val}' for key, val in params.items())}"
-
-    return thumbnail_url
-
-
-def fetch_wms(url: str, max_retries: int = 3, retry_delay: int = 1):
-    """
-    Function fetching an image from OGC server. The request is performed based on the WMS URL.
-    In case access_token in not present in the URL , and Geoserver is used and the OGC backend, Basic Authentication
-    is used instead. If image retrieval fails, function retries to fetch the image max_retries times, waiting
-    retry_delay seconds between consecutive requests.
-
-    :param url: WMS URL of the image
-    :param max_retries: maximum number of retries before skipping retrieval
-    :param retry_delay: number of seconds waited between retries
-    :returns: retrieved image
-    """
-
     # prepare authorization for WMS service
     headers = {}
-    if "access_token" not in url:
-        if url.startswith(settings.OGC_SERVER["default"]["LOCATION"]):
+    if "access_token" not in additional_kwargs.keys():
+        if thumbnail_url.startswith(settings.OGC_SERVER["default"]["LOCATION"]):
             # for the Geoserver backend, use Basic Auth, if access_token is not provided
             _user = settings.OGC_SERVER["default"].get("USER")
             _pwd = settings.OGC_SERVER["default"].get("PASSWORD")
             encoded_credentials = base64.b64encode(f"{_user}:{_pwd}".encode("UTF-8")).decode("ascii")
             headers["Authorization"] = f"Basic {encoded_credentials}"
 
-    image = None
+    wms = WebMapService(f"{thumbnail_url}{wms_endpoint}", version=wms_version, headers=headers)
 
+    image = None
     for retry in range(max_retries):
         try:
             # fetch data
-            resp, image = http_client.request(url, headers=headers)
+            image = wms.getmap(
```

Comment:
@bieganowski can we please check if this method allows a timeout parameter? It should as far as I remember.
In that case we should pass a timeout to the request, possibly taken from the OGC settings of GeoNode (similarly to what the http_client does).

Reply:
`timeout` added